Zend/zend_inheritance.c

```diff
@@ -173,72 +174,205 @@ static zend_always_inline zend_bool zend_iterable_compatibility_check(zend_arg_i
+			// todo: does ce1 == ce2 above catch all possible cases?
+			return 0;
+		case COVARIANT:
+			return instanceof_function(ce1, ce2);
+		case CONTRAVARIANT:
+			return instanceof_function(ce2, ce1);
 
-		fe_class_name = ZEND_TYPE_NAME(fe_arg_info->type);
-		class_name = ZSTR_VAL(fe_class_name);
-		class_name_len = ZSTR_LEN(fe_class_name);
-		if (class_name_len == sizeof("parent")-1 && !strcasecmp(class_name, "parent") && proto->common.scope) {
-			fe_class_name = zend_string_copy(proto->common.scope->name);
-		} else if (class_name_len == sizeof("self")-1 && !strcasecmp(class_name, "self") && fe->common.scope) {
-			fe_class_name = zend_string_copy(fe->common.scope->name);
-		} else {
-			zend_string_addref(fe_class_name);
-		}
+		EMPTY_SWITCH_DEFAULT_CASE();
+	}
+}
+
+/* This int return is not a boolean, but will often be treated this way.
+ *   *  0 means there is definitely an error.
+ *   *  1 means there are definitely not any errors.
+ *   * -1 means there are no known errors but it is not known to be good
+ *     either (there is not enough information at the moment).
+ */
+static
+int _check_inherited_arg_info(
+	const zend_function *fe, zend_arg_info *fe_arg_info,
+	const zend_function *proto, zend_arg_info *proto_arg_info,
+	_variance variance) /* {{{ */
+{
+	zend_type fe_type = fe_arg_info->type;
+	zend_type proto_type = proto_arg_info->type;
+	zend_long fe_type_code = ZEND_TYPE_CODE(fe_type);
+	zend_long proto_type_code = ZEND_TYPE_CODE(proto_type);
+
+	ZEND_ASSERT(ZEND_TYPE_IS_SET(fe_type));
+	ZEND_ASSERT(ZEND_TYPE_IS_SET(proto_type));
+
+	if (variance == COVARIANT && ZEND_TYPE_ALLOW_NULL(fe_type) && !ZEND_TYPE_ALLOW_NULL(proto_type)) {
+		return 0;
+	}
+
+	/* This introduces BC break described at https://bugs.php.net/bug.php?id=72119 */
+	if (variance == CONTRAVARIANT && ZEND_TYPE_ALLOW_NULL(proto_type) && !ZEND_TYPE_ALLOW_NULL(fe_type)) {
+		return 0;
+	}
+
+	if (ZEND_TYPE_IS_CLASS(fe_type)) {
+		zend_string *fe_class_name =
+			_resolve_parent_and_self(fe, ZEND_TYPE_NAME(fe_type));
+		int code = 1;
+		if (ZEND_TYPE_IS_CLASS(proto_type)) {
+			zend_string *proto_class_name =
+				_resolve_parent_and_self(proto, ZEND_TYPE_NAME(proto_type));
 
-		proto_class_name = ZEND_TYPE_NAME(proto_arg_info->type);
-		class_name = ZSTR_VAL(proto_class_name);
-		class_name_len = ZSTR_LEN(proto_class_name);
-		if (class_name_len == sizeof("parent")-1 && !strcasecmp(class_name, "parent") && proto->common.scope && proto->common.scope->parent) {
-			proto_class_name = zend_string_copy(proto->common.scope->parent->name);
-		} else if (class_name_len == sizeof("self")-1 && !strcasecmp(class_name, "self") && proto->common.scope) {
-			proto_class_name = zend_string_copy(proto->common.scope->name);
+			if (!zend_string_equals_ci(fe_class_name, proto_class_name)) {
+				if (fe->common.type == ZEND_USER_FUNCTION) {
+					zend_class_entry * fe_ce = zend_lookup_class(fe_class_name);
+					zend_class_entry * proto_ce = zend_lookup_class(proto_class_name);
```

Comment:
If I'm understanding this right, these calls can also happen during compilation. In that case it's necessary to respect flags like ZEND_COMPILE_IGNORE_OTHER_FILES to avoid introducing invalid dependencies in opcache cache state.

Reply:
Also nit: Extra space after `*`.

Reply:
I don't understand what it means to "respect flags like ZEND_COMPILE_IGNORE_OTHER_FILES to avoid introducing invalid dependencies in opcache cache state". Can you educate me?

Reply:
Opcache (ignoring preloading here) assumes that each file is independent. As such, it cannot use any information from other files, such as classes defined there, during compilation. This is what the ZEND_COMPILE_IGNORE_OTHER_FILES flag controls. Additionally, on Windows it may happen that an opcache state is reused with a different base state of internal classes. This is what ZEND_COMPILE_IGNORE_INTERNAL_CLASSES controls.

You can find an example of these flags in action here: https://github.com/php/php-src/blob/master/Zend/zend_compile.c#L6373

You'll probably want to add a helper that does the class lookup while respecting these flags (if lookup happens during compilation, this is not relevant at runtime).